046-analyse-broker-metrics-before-rolling-kafka.md

```diff
@@ -0,0 +1,63 @@
+
+KafkaAgent currently collects `kafka.server:type=KafkaServer,name=BrokerState` metric and creates a `kafka-ready` file on disk if the broker state value is `3`(`RUNNING`). The Kafka readiness check passes if it finds the file on disk.
+
+## Motivation 
+
+To address the issue described above, the KafkaRoller needs to be aware of the current broker state, so that if a broker is performing log recovery, then it waits longer instead of force restarting it.
+
+## Proposal
+
+The proposal is a REST API endpoint on the KafkaAgent which exposes the BrokerState metric that it already uses. And modify KafkaRoller so that it queries the new endpoint to find out what state a broker is in. 
+
+1. Add a web server to KafkaAgent that exposes an endpoint which returns a value of metric
+    - Configure Jetty server to only accept SSL connections on a port 8443.
+    - Expose an endpoint `/broker-state`.
+    - Add a servlet to handle a `GET` request and return an integer value of `BrokerState` metric.
+    - It will reuse broker's own certificate, which will already be in the container.
+2. Update KafkaRoller to query the broker state from the KafkaAgent API
+    - Currently KafkaRoller obtains headless service for a given pod and uses it to boostrap Kafka AdminClient. It can use the same mechanism to get the address for the KafkaAgent endpoint.
+    - Instead of checking broker's readiness through `isReady()` function, utilise the new restful endpoint to determine if the broker is ready.
+    - After restarting a broker if it does not become ready within the operational timeout period, check whether the broker is performing log recovery by querying the new endpoint:
+        - If log recovery is in progress, do not restart the broker.
+        - Otherwise, force restart the broker which is the current behabviour.
+    - Add a new configurable timeout for log recovery that is 60 minutes by default.
+    - If the log recovery does not complete within the timeout period, then fail the reconciliation and report an error for the human operator to investigate further. 
+
+### For future consideration:
+- The operator might need access to more broker metrics. We can easily implement additional endpoints to expose them.
+- The readiness check could also make use of the KafkaAgent endpoint to get the broker state instead of checking for a file on disk.
+- The liveness and readiness checks potentially will need to access non yammer metrics for KRAFT, which could be exposed in the same manner, but would need a new mechanism to collect them.
+
+## Affected/not affected projects
+
+* strimzi/kafka-cluster-operator
+
+## Compatibility
+
+When the operator gets updated with the KafkaRoller change, it will expect the KafkaAgent endpoint to be there. However, KafkaAgent would only get updated after the operator.
```

Comment:
```suggestion
When the operator gets updated with the KafkaRoller change, it will expect the KafkaAgent endpoint to be there. However, KafkaAgent would only get updated after the operator.

To maintain backwards compatibility, the following process is suggested:
```